### DISCUSSION TITLE
Actionable diagnostics / autofix

### DISCUSSION THREAD

**Discussion by kubukoz** [awslabs/smithy#2281] [Ideas]
When building a validator/linter, one might want to give the user an option to automatically apply a fix.
For example, consider a linter of unused shapes: it could offer the action to remove such shapes, or to add a suppression where appropriate.

Of course, this particular example is a built-in validator, and the removal of unused shapes could be special-cased [in the LSP](https://github.com/smithy-lang/smithy-language-server/issues/13). However, making it extensible would:

- allow Smithy users to implement rewrites together with their protocol, instead of the LSP acting as the centralized source of refactors
- allow applying the rewrites/refactors from other tools, such as the Smithy CLI, without having to launch the full LSP experience.
- help keep the rewrites consistent with the rules of such a validator, due to them living in the same project.

What do you think?

---

The implementation of a rewrite API is out of scope for this discussion: I'm thinking of a syntax tree-aware [Scalafix-like](https://scalacenter.github.io/scalafix/) thing, but for starters we could leave the details up to the implementor of the particular rewrite.

Other than rewrites, this API could be extended with navigation (e.g. "you used thing X here, but it conflicts with Y here"), but that can be a next step. There's a lot we could do with this!

This is inspired by [ideas](https://www.chris-kipp.io/blog/the-journey-of-a-dotty-diagnostic) from [the Scala ecosystem](https://contributors.scala-lang.org/t/roadmap-for-actionable-diagnostics/6172), but I think it originates from (among others) Rust.

![image](https://github.com/smithy-lang/smithy/assets/894884/9ece29a8-b996-4cf1-bc9b-140aac44aab1)

**Comment by mtdowling** [comment]
Refactoring support to fix issues would be great. It would be nice if it was a generic library, like you said, so it can be used in the CLI independent of the LSP. We'd like to eventually implement something like this on top of smithy-syntax + smithy-model (use smithy-model validation events and the token tree of smithy-syntax to detect refactorable issues, then use smithy-syntax to fix them). Scalafix and how they wire up to the LSP and CLI looks like a good prior art for inspiration.
